atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -728,19 +736,34 @@ private Comparator<Cell> preserveInputRowOrder(List<Map.Entry<Cell, Value>> inpu
      * per non-empty row.
      */
     private Iterator<Map.Entry<byte[], RowColumnRangeIterator>> partitionByRow(
-            Iterator<Map.Entry<Cell, Value>> rawResults) {
+            Iterator<Map.Entry<Cell, Value>> rawResults, Iterator<byte[]> expectedRows) {
         PeekingIterator<Map.Entry<Cell, Value>> peekableRawResults = Iterators.peekingIterator(rawResults);
-        return new AbstractIterator<Map.Entry<byte[], RowColumnRangeIterator>>() {
+        return new AbstractIterator<>() {
             byte[] prevRowName;
 
             @Override
             protected Map.Entry<byte[], RowColumnRangeIterator> computeNext() {
                 finishConsumingPreviousRow(peekableRawResults);
-                if (!peekableRawResults.hasNext()) {
+                if (!expectedRows.hasNext()) {
+                    if (peekableRawResults.hasNext()) {
+                        throw new SafeIllegalStateException(
+                                "Iterators are not consistent: there is some data returned by getRowsColumnRange()"
+                                        + " even when we expect no more rows. This is likely an AtlasDB product bug.",
+                                UnsafeArg.of("nextRawResult", peekableRawResults.peek()));
+                    }
                     return endOfData();
                 }
+                byte[] nextExpectedRow = expectedRows.next();
+
+                if (!peekableRawResults.hasNext()) {
+                    return Maps.immutableEntry(nextExpectedRow, EmptyRowColumnRangeIterator.INSTANCE);
+                }
                 byte[] nextRowName = peekableRawResults.peek().getKey().getRowName();
-                Iterator<Map.Entry<Cell, Value>> columnsIterator = new AbstractIterator<Map.Entry<Cell, Value>>() {
+                if (!Arrays.equals(nextRowName, nextExpectedRow)) {
```

Comment:
Can we always operate off the assumption that both our iterators will be in the same order (whether, ascending, descending)? Otherwise a nasty bug could manifest here, in that, we return an empty row, when there really should be data.

Reply:
(I see that `getRowColumnRangePostFiltered` uses a SortedMap to maintain input order, but is still a good question more broadly)

Reply:
Yeah this does rely on both matching iteration order. This _does_ hold, since the iterator passed in is a filtered version of the raw results returned from a KVS call to `getRowsColumnRange`, and the API of that defines ordering to follow the input iterable of rows passed in:

```
     * @return a {@link RowColumnRangeIterator} that can iterate over all the retrieved values. Results for different
     *         rows are in the same order as they are provided in {@code rows}. All columns for a given row are adjacent
     *         and sorted by increasing column name.
```

There's a valid question on if/how we could make this more defensive. I was thinking some kind of wrapper type around the iterators? Though not sure if it's worth the complexity

Reply:
Yeah I'd say that current implementation this is good, but just trying to think how to future proof this method in particular to prevent against future refactors/changes etc. It does appear that `testGetRowColumnRangeMultipleRows` in `AbstractKeyValueServiceTest`, so maybe that's sufficient? 

Only other idea is that in testing we validate sort order when we call this method (within the method itself), but everywhere else we do not. This change is quite ugly though.

Reply:
So yeah, the AbstractKVSTest would help us validate that the KVS layer indeed satisfies its API. There is also a layer in between that is how we manipulate the iterators returned from the KVS, so I wouldn't see that alone as sufficient.

This is called by getRowsColumnRange 4 arg version, so I _think_ the test `getRowsColumnRangeIncludesRowsWithOnlyLocalWritesCorrectly` in `SnapshotTransactionTest` will almost certainly end up failing if we get this wrong. I've also added another test that tests the "normal" case with all the rows already in the KVS. The combination of these tests with the SnapTrans one would be sufficient for me at least.